File: backtestApp/panels/strategy_specific_panels/buy_heikin_green_panel.py
from PyQt5.QtWidgets import (QGroupBox, QVBoxLayout, QGridLayout, QCheckBox, 
                           QSpinBox, QLabel, QFrame, QHBoxLayout)
from ..base_panel import BasePanel

class BuyHeikinGreenPanel(BasePanel):
    """Panel spécifique à la stratégie BuyHeikinGreen."""
    
    def __init__(self, parent=None):
        super().__init__(parent)
    
    def create(self):
        """Crée le panel des paramètres spécifiques à BuyHeikinGreen."""
        strategy_group = QGroupBox("Paramètres BuyHeikinGreen")
        strategy_layout = QVBoxLayout()
        
        # Section EMA Court
        ema_short_group = QGroupBox("EMA Court")
        ema_short_layout = QVBoxLayout()
        
        # Checkbox pour activer/désactiver le filtre EMA court
        self.widgets['ema_short_filter_check'] = QCheckBox("Activer le filtre EMA court")
        self.widgets['ema_short_filter_check'].setChecked(True)
        ema_short_layout.addWidget(self.widgets['ema_short_filter_check'])
        
        # Frame pour les paramètres de l'EMA court
        ema_short_params = QFrame()
        ema_short_params_layout = QHBoxLayout()
        ema_short_params_layout.addWidget(QLabel("Période:"))
        self.widgets['ema_short_spin'] = QSpinBox()
        self.widgets['ema_short_spin'].setRange(1, 500)
        self.widgets['ema_short_spin'].setValue(20)
        ema_short_params_layout.addWidget(self.widgets['ema_short_spin'])
        ema_short_params_layout.addStretch()
        ema_short_params.setLayout(ema_short_params_layout)
        ema_short_layout.addWidget(ema_short_params)
        
        ema_short_group.setLayout(ema_short_layout)
        strategy_layout.addWidget(ema_short_group)
        
        # Section EMA Long
        ema_long_group = QGroupBox("EMA Long")
        ema_long_layout = QVBoxLayout()
        
        # Checkbox pour activer/désactiver le filtre EMA long
        self.widgets['ema_long_filter_check'] = QCheckBox("Activer le filtre EMA long")
        self.widgets['ema_long_filter_check'].setChecked(True)
        ema_long_layout.addWidget(self.widgets['ema_long_filter_check'])
        
        # Frame pour les paramètres de l'EMA long
        ema_long_params = QFrame()
        ema_long_params_layout = QHBoxLayout()
        ema_long_params_layout.addWidget(QLabel("Période:"))
        self.widgets['ema_long_spin'] = QSpinBox()
        self.widgets['ema_long_spin'].setRange(1, 500)
        self.widgets['ema_long_spin'].setValue(200)
        ema_long_params_layout.addWidget(self.widgets['ema_long_spin'])
        ema_long_params_layout.addStretch()
        ema_long_params.setLayout(ema_long_params_layout)
        ema_long_layout.addWidget(ema_long_params)
        
        ema_long_group.setLayout(ema_long_layout)
        strategy_layout.addWidget(ema_long_group)
        
        # Section Stochastique
        stoch_group = QGroupBox("Stochastique")
        stoch_layout = QVBoxLayout()
        
        # Checkbox pour activer/désactiver le filtre Stochastique
        self.widgets['stoch_filter_check'] = QCheckBox("Activer le filtre Stochastique")
        self.widgets['stoch_filter_check'].setChecked(True)
        stoch_layout.addWidget(self.widgets['stoch_filter_check'])
        
        # Frame pour les paramètres du Stochastique
        stoch_params = QFrame()
        stoch_params_layout = QGridLayout()
        
        stoch_params_layout.addWidget(QLabel("Fast %K:"), 0, 0)
        self.widgets['fastk_spin'] = QSpinBox()
        self.widgets['fastk_spin'].setRange(1, 100)
        self.widgets['fastk_spin'].setValue(10)
        stoch_params_layout.addWidget(self.widgets['fastk_spin'], 0, 1)
        
        stoch_params_layout.addWidget(QLabel("Slow %K:"), 1, 0)
        self.widgets['slowk_spin'] = QSpinBox()
        self.widgets['slowk_spin'].setRange(1, 100)
        self.widgets['slowk_spin'].setValue(10)
        stoch_params_layout.addWidget(self.widgets['slowk_spin'], 1, 1)
        
        stoch_params_layout.addWidget(QLabel("Slow %D:"), 2, 0)
        self.widgets['slowd_spin'] = QSpinBox()
        self.widgets['slowd_spin'].setRange(1, 100)
        self.widgets['slowd_spin'].setValue(3)
        stoch_params_layout.addWidget(self.widgets['slowd_spin'], 2, 1)
        
        stoch_params_layout.addWidget(QLabel("Seuil filtre %K:"), 3, 0)
        self.widgets['stoch_threshold_spin'] = QSpinBox()
        self.widgets['stoch_threshold_spin'].setRange(1, 99)
        self.widgets['stoch_threshold_spin'].setValue(50)
        stoch_params_layout.addWidget(self.widgets['stoch_threshold_spin'], 3, 1)
        
        stoch_params.setLayout(stoch_params_layout)
        stoch_layout.addWidget(stoch_params)
        
        stoch_group.setLayout(stoch_layout)
        strategy_layout.addWidget(stoch_group)
        
        # Section pour le filtre de bougie Heikin Ashi précédente rouge
        ha_group = QGroupBox("Condition bougie Heikin Ashi")
        ha_layout = QVBoxLayout()
        
        self.widgets['previous_ha_candle_red_check'] = QCheckBox("Activer le filtre bougie précédente rouge")
        self.widgets['previous_ha_candle_red_check'].setChecked(True)
        ha_layout.addWidget(self.widgets['previous_ha_candle_red_check'])
        
        ha_group.setLayout(ha_layout)
        strategy_layout.addWidget(ha_group)
        
        # Connexion des signaux pour activer/désactiver les widgets en fonction des checkboxes
        self.widgets['ema_short_filter_check'].toggled.connect(
            lambda checked: self._toggle_widget_group([self.widgets['ema_short_spin']], checked))
        
        self.widgets['ema_long_filter_check'].toggled.connect(
            lambda checked: self._toggle_widget_group([self.widgets['ema_long_spin']], checked))
        
        self.widgets['stoch_filter_check'].toggled.connect(
            lambda checked: self._toggle_widget_group([
                self.widgets['fastk_spin'], 
                self.widgets['slowk_spin'], 
                self.widgets['slowd_spin'], 
                self.widgets['stoch_threshold_spin']
            ], checked))
        
        strategy_group.setLayout(strategy_layout)
        return strategy_group
    
    def _toggle_widget_group(self, widgets, enabled):
        """Active/désactive un groupe de widgets."""
        for widget in widgets:
            widget.setEnabled(enabled)
    
    def get_values(self):
        """Récupère les valeurs des widgets du panel."""
        #TODO : Il serait bien ici de return directement l'objet de la stratégie : le dataclass : BuyHeikinGreenConfig
        return {
            # Périodes des EMA
            'ema_short_period': self.widgets['ema_short_spin'].value(),
            'ema_long_period': self.widgets['ema_long_spin'].value(),
            
            # Paramètres Stochastique
            'stoch_fastk': self.widgets['fastk_spin'].value(),
            'stoch_slowk': self.widgets['slowk_spin'].value(),
            'stoch_slowd': self.widgets['slowd_spin'].value(),
            'stoch_threshold': self.widgets['stoch_threshold_spin'].value(),
            
            # Activation des filtres
            'use_ema_short_filter': self.widgets['ema_short_filter_check'].isChecked(),
            'use_ema_long_filter': self.widgets['ema_long_filter_check'].isChecked(),
            'use_stoch_filter': self.widgets['stoch_filter_check'].isChecked(),
            'use_previous_ha_candle_red_filter': self.widgets['previous_ha_candle_red_check'].isChecked(),
        }
    
    def get_required_indicators(self):
        """
        Retourne les indicateurs nécessaires pour la stratégie BuyHeikinGreen.
        Cette méthode est utilisée par l'application de backtest pour précalculer les indicateurs.
        
        Returns:
            dict: Un dictionnaire des indicateurs à précalculer au format:
                {'NOM_INDICATEUR': [[param1, param2, ...], [autre_config], ...]}
        """
        indicators = {}
        
        # EMA
        ema_periods = []
        if self.widgets['ema_short_filter_check'].isChecked():
            ema_periods.append([self.widgets['ema_short_spin'].value()])
        
        if self.widgets['ema_long_filter_check'].isChecked():
            ema_periods.append([self.widgets['ema_long_spin'].value()])
        
        # Ajouter une EMA moyenne pour l'affichage si elle existe
        if hasattr(self.widgets, 'ema_medium_check') and self.widgets.get('ema_medium_check') and self.widgets['ema_medium_check'].isChecked():
            ema_periods.append([self.widgets['ema_medium_spin'].value()])
        
        if ema_periods:
            indicators['EMA'] = ema_periods
        
        # Stochastique
        if self.widgets['stoch_filter_check'].isChecked():
            indicators['STOCH'] = [[
                self.widgets['fastk_spin'].value(),
                self.widgets['slowk_spin'].value(),
                self.widgets['slowd_spin'].value()
            ]]
        
        # ATR (pour le calcul des stops si nécessaire)
        if hasattr(self.widgets, 'atr_check') and self.widgets.get('atr_check') and self.widgets['atr_check'].isChecked():
            indicators['ATR'] = [[self.widgets['atr_period_spin'].value()]]
        
        return indicators
    

File: cpp_backtestApp/src/panels/strategy_specific_panels/buy_heikin_green_panel.cpp
#include "panels/strategy_specific_panels/buy_heikin_green_panel.h"
#include <QDebug>

BuyHeikinGreenPanel::BuyHeikinGreenPanel(QWidget* parent)
    : QObject(parent), BasePanel(parent)
{
}

QGroupBox* BuyHeikinGreenPanel::create()
{
    QGroupBox* strategyGroup = new QGroupBox("Paramètres BuyHeikinGreen");
    QVBoxLayout* strategyLayout = new QVBoxLayout();
    
    // Section EMA Court
    QGroupBox* emaShortGroup = new QGroupBox("EMA Court");
    QVBoxLayout* emaShortLayout = new QVBoxLayout();
    
    // Checkbox pour activer/désactiver le filtre EMA court
    m_widgets["ema_short_filter_check"] = new QCheckBox("Activer le filtre EMA court");
    static_cast<QCheckBox*>(m_widgets["ema_short_filter_check"])->setChecked(true);
    connect(static_cast<QCheckBox*>(m_widgets["ema_short_filter_check"]), &QCheckBox::toggled,
            [this](bool checked) {
                _toggleWidgetGroup({"ema_short_spin"}, checked);
            });
    emaShortLayout->addWidget(m_widgets["ema_short_filter_check"]);
    
    // Frame pour les paramètres de l'EMA court
    QFrame* emaShortParams = new QFrame();
    QHBoxLayout* emaShortParamsLayout = new QHBoxLayout();
    emaShortParamsLayout->addWidget(new QLabel("Période:"));
    m_widgets["ema_short_spin"] = new QSpinBox();
    static_cast<QSpinBox*>(m_widgets["ema_short_spin"])->setRange(1, 500);
    static_cast<QSpinBox*>(m_widgets["ema_short_spin"])->setValue(20);
    emaShortParamsLayout->addWidget(m_widgets["ema_short_spin"]);
    emaShortParamsLayout->addStretch();
    emaShortParams->setLayout(emaShortParamsLayout);
    emaShortLayout->addWidget(emaShortParams);
    
    emaShortGroup->setLayout(emaShortLayout);
    strategyLayout->addWidget(emaShortGroup);
    
    // Section EMA Long
    QGroupBox* emaLongGroup = new QGroupBox("EMA Long");
    QVBoxLayout* emaLongLayout = new QVBoxLayout();
    
    // Checkbox pour activer/désactiver le filtre EMA long
    m_widgets["ema_long_filter_check"] = new QCheckBox("Activer le filtre EMA long");
    static_cast<QCheckBox*>(m_widgets["ema_long_filter_check"])->setChecked(true);
    connect(static_cast<QCheckBox*>(m_widgets["ema_long_filter_check"]), &QCheckBox::toggled,
            [this](bool checked) {
                _toggleWidgetGroup({"ema_long_spin"}, checked);
            });
    emaLongLayout->addWidget(m_widgets["ema_long_filter_check"]);
    
    // Frame pour les paramètres de l'EMA long
    QFrame* emaLongParams = new QFrame();
    QHBoxLayout* emaLongParamsLayout = new QHBoxLayout();
    emaLongParamsLayout->addWidget(new QLabel("Période:"));
    m_widgets["ema_long_spin"] = new QSpinBox();
    static_cast<QSpinBox*>(m_widgets["ema_long_spin"])->setRange(1, 500);
    static_cast<QSpinBox*>(m_widgets["ema_long_spin"])->setValue(200);
    emaLongParamsLayout->addWidget(m_widgets["ema_long_spin"]);
    emaLongParamsLayout->addStretch();
    emaLongParams->setLayout(emaLongParamsLayout);
    emaLongLayout->addWidget(emaLongParams);
    
    emaLongGroup->setLayout(emaLongLayout);
    strategyLayout->addWidget(emaLongGroup);
    
    // Section RSI
    QGroupBox* rsiGroup = new QGroupBox("RSI");
    QVBoxLayout* rsiLayout = new QVBoxLayout();
    
    // Checkbox pour activer/désactiver le filtre RSI
    m_widgets["rsi_filter_check"] = new QCheckBox("Activer le filtre RSI");
    static_cast<QCheckBox*>(m_widgets["rsi_filter_check"])->setChecked(true);
    connect(static_cast<QCheckBox*>(m_widgets["rsi_filter_check"]), &QCheckBox::toggled,
            [this](bool checked) {
                _toggleWidgetGroup({"rsi_period_spin", "rsi_threshold_spin"}, checked);
            });
    rsiLayout->addWidget(m_widgets["rsi_filter_check"]);
    
    // Frame pour les paramètres du RSI
    QFrame* rsiParams = new QFrame();
    QGridLayout* rsiParamsLayout = new QGridLayout();
    
    rsiParamsLayout->addWidget(new QLabel("Période:"), 0, 0);
    m_widgets["rsi_period_spin"] = new QSpinBox();
    static_cast<QSpinBox*>(m_widgets["rsi_period_spin"])->setRange(2, 100);
    static_cast<QSpinBox*>(m_widgets["rsi_period_spin"])->setValue(14);
    rsiParamsLayout->addWidget(m_widgets["rsi_period_spin"], 0, 1);
    
    rsiParamsLayout->addWidget(new QLabel("Seuil filtre:"), 1, 0);
    m_widgets["rsi_threshold_spin"] = new QSpinBox();
    static_cast<QSpinBox*>(m_widgets["rsi_threshold_spin"])->setRange(1, 99);
    static_cast<QSpinBox*>(m_widgets["rsi_threshold_spin"])->setValue(30);
    rsiParamsLayout->addWidget(m_widgets["rsi_threshold_spin"], 1, 1);
    
    rsiParams->setLayout(rsiParamsLayout);
    rsiLayout->addWidget(rsiParams);
    
    rsiGroup->setLayout(rsiLayout);
    strategyLayout->addWidget(rsiGroup);
    
    // Section Stochastique
    QGroupBox* stochGroup = new QGroupBox("Stochastique");
    QVBoxLayout* stochLayout = new QVBoxLayout();
    
    // Checkbox pour activer/désactiver le filtre Stochastique
    m_widgets["stoch_filter_check"] = new QCheckBox("Activer le filtre Stochastique");
    static_cast<QCheckBox*>(m_widgets["stoch_filter_check"])->setChecked(true);
    connect(static_cast<QCheckBox*>(m_widgets["stoch_filter_check"]), &QCheckBox::toggled,
            [this](bool checked) {
                _toggleWidgetGroup({"fastk_spin", "slowk_spin", "slowd_spin", "stoch_threshold_spin"}, checked);
            });
    stochLayout->addWidget(m_widgets["stoch_filter_check"]);
    
    // Frame pour les paramètres du Stochastique
    QFrame* stochParams = new QFrame();
    QGridLayout* stochParamsLayout = new QGridLayout();
    
    stochParamsLayout->addWidget(new QLabel("Fast %K:"), 0, 0);
    m_widgets["fastk_spin"] = new QSpinBox();
    static_cast<QSpinBox*>(m_widgets["fastk_spin"])->setRange(1, 100);
    static_cast<QSpinBox*>(m_widgets["fastk_spin"])->setValue(10);
    stochParamsLayout->addWidget(m_widgets["fastk_spin"], 0, 1);
    
    stochParamsLayout->addWidget(new QLabel("Slow %K:"), 1, 0);
    m_widgets["slowk_spin"] = new QSpinBox();
    static_cast<QSpinBox*>(m_widgets["slowk_spin"])->setRange(1, 100);
    static_cast<QSpinBox*>(m_widgets["slowk_spin"])->setValue(7);
    stochParamsLayout->addWidget(m_widgets["slowk_spin"], 1, 1);
    
    stochParamsLayout->addWidget(new QLabel("Slow %D:"), 2, 0);
    m_widgets["slowd_spin"] = new QSpinBox();
    static_cast<QSpinBox*>(m_widgets["slowd_spin"])->setRange(1, 100);
    static_cast<QSpinBox*>(m_widgets["slowd_spin"])->setValue(3);
    stochParamsLayout->addWidget(m_widgets["slowd_spin"], 2, 1);
    
    stochParamsLayout->addWidget(new QLabel("Seuil:"), 3, 0);
    m_widgets["stoch_threshold_spin"] = new QSpinBox();
    static_cast<QSpinBox*>(m_widgets["stoch_threshold_spin"])->setRange(1, 99);
    static_cast<QSpinBox*>(m_widgets["stoch_threshold_spin"])->setValue(20);
    stochParamsLayout->addWidget(m_widgets["stoch_threshold_spin"], 3, 1);
    
    stochParams->setLayout(stochParamsLayout);
    stochLayout->addWidget(stochParams);
    
    stochGroup->setLayout(stochLayout);
    strategyLayout->addWidget(stochGroup);
    
    // Section Filtre Bougie Précédente
    QGroupBox* candleFilterGroup = new QGroupBox("Filtre de bougie");
    QVBoxLayout* candleFilterLayout = new QVBoxLayout();
    
    m_widgets["previous_ha_candle_red_filter_check"] = new QCheckBox("La bougie précédente doit être rouge");
    candleFilterLayout->addWidget(m_widgets["previous_ha_candle_red_filter_check"]);
    
    candleFilterGroup->setLayout(candleFilterLayout);
    strategyLayout->addWidget(candleFilterGroup);
    
    // Ajouter un espace extensible en bas
    strategyLayout->addStretch();
    
    strategyGroup->setLayout(strategyLayout);
    return strategyGroup;
}

QMap<QString, QVariant> BuyHeikinGreenPanel::getValues()
{
    QMap<QString, QVariant> values;
    
    // EMA Court
    values["use_ema_short_filter"] = static_cast<QCheckBox*>(m_widgets["ema_short_filter_check"])->isChecked();
    values["ema_short_period"] = static_cast<QSpinBox*>(m_widgets["ema_short_spin"])->value();
    
    // EMA Long
    values["use_ema_long_filter"] = static_cast<QCheckBox*>(m_widgets["ema_long_filter_check"])->isChecked();
    values["ema_long_period"] = static_cast<QSpinBox*>(m_widgets["ema_long_spin"])->value();
    
    // RSI
    values["use_rsi_filter"] = static_cast<QCheckBox*>(m_widgets["rsi_filter_check"])->isChecked();
    values["rsi_period"] = static_cast<QSpinBox*>(m_widgets["rsi_period_spin"])->value();
    values["rsi_threshold"] = static_cast<QSpinBox*>(m_widgets["rsi_threshold_spin"])->value();
    
    // Stochastique
    values["use_stoch_filter"] = static_cast<QCheckBox*>(m_widgets["stoch_filter_check"])->isChecked();
    values["stoch_fastk"] = static_cast<QSpinBox*>(m_widgets["fastk_spin"])->value();
    values["stoch_slowk"] = static_cast<QSpinBox*>(m_widgets["slowk_spin"])->value();
    values["stoch_slowd"] = static_cast<QSpinBox*>(m_widgets["slowd_spin"])->value();
    values["stoch_threshold"] = static_cast<QSpinBox*>(m_widgets["stoch_threshold_spin"])->value();
    
    // Filtre bougie
    values["use_previous_ha_candle_red_filter"] = static_cast<QCheckBox*>(m_widgets["previous_ha_candle_red_filter_check"])->isChecked();
    
    return values;
}

void BuyHeikinGreenPanel::setValues(const QMap<QString, QVariant>& values)
{
    // EMA Court
    if (values.contains("use_ema_short_filter")) {
        static_cast<QCheckBox*>(m_widgets["ema_short_filter_check"])->setChecked(values["use_ema_short_filter"].toBool());
    }
    
    if (values.contains("ema_short_period")) {
        static_cast<QSpinBox*>(m_widgets["ema_short_spin"])->setValue(values["ema_short_period"].toInt());
    }
    
    // EMA Long
    if (values.contains("use_ema_long_filter")) {
        static_cast<QCheckBox*>(m_widgets["ema_long_filter_check"])->setChecked(values["use_ema_long_filter"].toBool());
    }
    
    if (values.contains("ema_long_period")) {
        static_cast<QSpinBox*>(m_widgets["ema_long_spin"])->setValue(values["ema_long_period"].toInt());
    }
    
    // RSI
    if (values.contains("use_rsi_filter")) {
        static_cast<QCheckBox*>(m_widgets["rsi_filter_check"])->setChecked(values["use_rsi_filter"].toBool());
    }
    
    if (values.contains("rsi_period")) {
        static_cast<QSpinBox*>(m_widgets["rsi_period_spin"])->setValue(values["rsi_period"].toInt());
    }
    
    if (values.contains("rsi_threshold")) {
        static_cast<QSpinBox*>(m_widgets["rsi_threshold_spin"])->setValue(values["rsi_threshold"].toInt());
    }
    
    // Stochastique
    if (values.contains("use_stoch_filter")) {
        static_cast<QCheckBox*>(m_widgets["stoch_filter_check"])->setChecked(values["use_stoch_filter"].toBool());
    }
    
    if (values.contains("stoch_fastk")) {
        static_cast<QSpinBox*>(m_widgets["fastk_spin"])->setValue(values["stoch_fastk"].toInt());
    }
    
    if (values.contains("stoch_slowk")) {
        static_cast<QSpinBox*>(m_widgets["slowk_spin"])->setValue(values["stoch_slowk"].toInt());
    }
    
    if (values.contains("stoch_slowd")) {
        static_cast<QSpinBox*>(m_widgets["slowd_spin"])->setValue(values["stoch_slowd"].toInt());
    }
    
    if (values.contains("stoch_threshold")) {
        static_cast<QSpinBox*>(m_widgets["stoch_threshold_spin"])->setValue(values["stoch_threshold"].toInt());
    }
    
    // Filtre bougie
    if (values.contains("use_previous_ha_candle_red_filter")) {
        static_cast<QCheckBox*>(m_widgets["previous_ha_candle_red_filter_check"])->setChecked(
            values["use_previous_ha_candle_red_filter"].toBool());
    }
    
    // Mettre à jour l'état d'activation des widgets en fonction des checkboxes
    _toggleWidgetGroup({"ema_short_spin"}, 
                     static_cast<QCheckBox*>(m_widgets["ema_short_filter_check"])->isChecked());
    
    _toggleWidgetGroup({"ema_long_spin"}, 
                     static_cast<QCheckBox*>(m_widgets["ema_long_filter_check"])->isChecked());
    
    _toggleWidgetGroup({"rsi_period_spin", "rsi_threshold_spin"}, 
                     static_cast<QCheckBox*>(m_widgets["rsi_filter_check"])->isChecked());
    
    _toggleWidgetGroup({"fastk_spin", "slowk_spin", "slowd_spin", "stoch_threshold_spin"}, 
                     static_cast<QCheckBox*>(m_widgets["stoch_filter_check"])->isChecked());
}

void BuyHeikinGreenPanel::_toggleWidgetGroup(const QStringList& widgets, bool enabled)
{
    for (const QString& widgetName : widgets) {
        if (m_widgets.contains(widgetName)) {
            m_widgets[widgetName]->setEnabled(enabled);
        }
    }
}